File: 2025/07 July/31 Powerful Integer/main.py
# Link: https://www.geeksforgeeks.org/problems/powerfull-integer--170647/1


class Solution:
    def powerfulInteger(self, intervals, k):
        mpp = {}

        for start, end in intervals:
            mpp[start] = mpp.get(start, 0) + 1
            mpp[end + 1] = mpp.get(end + 1, 0) - 1

        ans = -1
        temp = 0

        for point in sorted(mpp):
            delta = mpp[point]

            if delta >= 0:
                temp += delta
                if temp >= k:
                    ans = point
            else:
                if temp >= k:
                    ans = point - 1
                temp += delta

        return ans


if __name__ == "__main__":
    sol = Solution()
    intervals = [[1, 3], [4, 6], [3, 4]]
    k = 2
    result = sol.powerfulInteger(intervals, k)
    print(f"Powerful Integer: {result}")

    intervals = [[1, 4], [12, 45], [3, 8], [10, 12]]
    k = 3
    result = sol.powerfulInteger(intervals, k)
    print(f"Powerful Integer: {result}")

    intervals = [[16, 21], [5, 8], [12, 17], [17, 29], [9, 24]]
    k = 3
    result = sol.powerfulInteger(intervals, k)
    print(f"Powerful Integer: {result}")

File: 2025/07 July/31 Powerful Integer/main.cpp
// Link: https://www.geeksforgeeks.org/problems/powerfull-integer--170647/1

#include <iostream>
#include <vector>
#include <map>
using namespace std;

class Solution
{
public:
    int powerfulInteger(vector<vector<int>> &intervals, int k)
    {
        map<int, int> mpp;

        for (auto &x : intervals)
        {
            mpp[x[0]] += 1;
            mpp[x[1] + 1] -= 1;
        }

        int ans = -1;
        int temp = 0;

        for (auto &x : mpp)
        {
            if (mpp[x.first] >= 0)
            {
                temp += mpp[x.first];
                if (temp >= k)
                {
                    ans = x.first;
                }
            }
            else
            {
                if (temp >= k)
                {
                    ans = x.first - 1;
                }
                temp += mpp[x.first];
            }
        }

        return ans;
    }
};

int main()
{
    Solution sol;
    vector<vector<int>> intervals = {{1, 3}, {4, 6}, {3, 4}};
    int k = 2;
    int result = sol.powerfulInteger(intervals, k);
    cout << "Powerful Integer: " << result << endl;

    intervals = {{1, 4}, {12, 45}, {3, 8}, {10, 12}};
    k = 3;
    result = sol.powerfulInteger(intervals, k);
    cout << "Powerful Integer: " << result << endl;

    intervals = {{16, 21}, {5, 8}, {12, 17}, {17, 29}, {9, 24}};
    k = 3;
    result = sol.powerfulInteger(intervals, k);
    cout << "Powerful Integer: " << result << endl;
    return 0;
}